awslabs/aws-lambda-powertools-typescript
# test(idempotency): add e2e tests for idempotency

## Description of your changes

This PR introduces e2e tests for idempotency using decorator and function wrapper. Several changes were also introduced to the interface and options type, here is the breakdown. 

### `IdempotencyOptions`

First, we have unwrapped the options object before passing it to the `IdempotencyHandler`, this helps to keep it generic and decide which parameters we should pass from decorator or wrapper. We also expanded the type to contain `dataKeywordArgument` and `config` as optional parameters. This enables us to apply decorator to lambda handler or any other method. In addition, we can use the same interface for decorator and wrapper. 


### Test cases

The test cases should help to test basic functionality. We have much more possible tests with permutations of input, config, persistence layer options and sdk options. Not all of them are covered yet and I suggest to use the beta phase to add more tests down the road. The current test cases are:

* decorator
  * call method twice sequentially, the response value is in ddb
  * call method twice parallel, the response value is in ddb and we got one error during second call
  * call method once with custom persistence layer options
  * call method that fails, there is no ddb entry
  * call method with keyword argument with two different key/value pairs, the response values are in ddb
* wrapper
  * call method twice sequentially, the response value is in ddb 
  * call method with custom persistence layer options



### Related issues, RFCs


**Issue number:** #447 

## Checklist


### Breaking change checklist

***Is it a breaking change?:*** NO


---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.
